File: java/MaxCounters.java
//https://codility.com/demo/results/training4SCUFZ-B5W/
/*
Please use the slider on the page above to see why you were wrong at the first place!
*/
class SolutionMaxCounter {
    public int[] solution(int N, int[] A) {
        int maxhere = 0, maxsofar = 0;
        int[] ans = new int[N];
        for(int i : A) {
            if(i == N + 1)
                maxsofar = maxhere;
            else {
                if(ans[i - 1] < maxsofar) ans[i - 1] = maxsofar + 1;
                else ans[i - 1] += 1;
                maxhere = Math.max(maxhere, ans[i - 1]);
            }
        }
        if(maxsofar > 0)
            for(int i = 0; i < N; ++i)
                if(ans[i] < maxsofar)
                    ans[i] = maxsofar;
        return ans;
    }
}

File: python/MaxCounters.py
#https://codility.com/demo/results/trainingNE5FFD-8AF/
def solution(N, A):
    ans = [0] * N
    maxAtHere, maxSofar = 0, 0
    for i in A:
        if i == N + 1:
            maxSofar = max(maxSofar, maxAtHere)
        else:
            ans[i - 1] = ans[i - 1] + 1 if ans[i - 1] >= maxSofar else maxSofar + 1
            maxAtHere = max(ans[i - 1], maxAtHere)
    for i in xrange(N):
        if ans[i] < maxSofar:
            ans[i] = maxSofar
    return ans
